Support of FRM302 module.

qba667 opened the pull request:
According to arrangement with Andre, first version of AFHDS3 protocol for OpenTX 2.3.
I will send the module to Andre so test on real hardware will be possible.
Currently tested only on X10 Platform. FlySky is running own tests meantime.
Builds for testing attached.
Firmware for the FRM302 - USART at 115200 - so all OpenTX radios can use it.
As already discussed with FlySky in next version communication will be changed to half duplex on SPORT. 

[Update _fix firmware FRM302_115200.zip](https://github.com/opentx/opentx/files/5361029/Update._fix.firmware.FRM302_115200.zip)

qba667 committed:
FRM302 stub implemented.

qba667 committed:
UI for X10 implemented for AFHDS3

qba667 committed:
First draft of FRM302 implementation according to Bertrand requirements.

qba667 committed:
AFHDS3 running at 115200

qba667 committed:
Non inverted telemetry handling with different baud rated implemented.
AFHDS 3 pulse generation improved.

qba667 committed:
AFHS3 simplified sending of frame.
Externa module code for AFHDS3 implemented for Teranis.

qba667 committed:
Debug code removed.
Polarity fixed for AFHDS3.

qba667 committed:
Not runing AFHDS3 as sync mode.

qba667 committed:
getModuleMode and setModuleMode can be used without including pulses.h
Using 115200 bps for AFHDS3.
Binding implemented for AFHDS3.

qba667 committed:
Fixed communication for FRM302.
Fixed binding for FRM302.

qba667 committed:
AFHDS3 merged with latest NV14 changes

qba667 committed:
Frm302 telemetry fixed.
Afhds3 sending periodic commands fixed.

qba667 committed:
Tarains implementation of FRM302 added.

qba667 committed:
AFHDS added for LCD 128x64

qba667 committed:
Fixed actual power display for FRM302

qba667 committed:
Merge remote-tracking branch 'origin/2.3' into FRM302

qba667 committed:
AFHDS3 cleanup - separation from multi.

qba667 committed:
Bind menu added for AFHDS3

qba667 committed:
Fixed power status for afhds3.
UI for X10 improved.

qba667 committed:
Fixed initial power of afhds3 module.

qba667 committed:
Afhds3 namespace fixed.

qba667 committed:
Reverted invalid commit - python version specific.

qba667 committed:
Merge branch '2.3' of https://github.com/opentx/opentx into FRM302

qba667 committed:
Reset of AFhds3 module impoved.

qba667 committed:
Afhds3 reset fixed.

qba667 committed:
Translations updated.

qba667 committed:
Fixed build on X9 platform.

qba667 committed:
Better PXX2 check implemented - fix compile errors.

qba667 committed:
Fixed compilation on sky9x.

qba667 committed:
Build fix for sky9x

qba667 committed:
Compile fix for sky9x.

qba667 committed:
AFHDS3 excluded for Sky9x baords - due to linking errors.

qba667 committed:
Sky9X afhds removed - not inverted serial implementation missing.

3djc commented:
It does not pass Travis, so we will be reviewing when it does :)

qba667 committed:
Compilation fixes for Sky9x.

qba667 committed:
Merge branch '2.3' of https://github.com/opentx/opentx into FRM302

qba667 committed:
FRM304 updated to latest version from NV14.

qba667 commented:
@3djc sky9x compile errors fixed.

3djc committed:
Compile fix

3djc commented:
X7:
![image](https://user-images.githubusercontent.com/5167938/80359818-ad1f7c00-887e-11ea-8b67-5ee795bcac6a.png)

should be

3djc commented:
I fixed 212 not compiling with AFHDS3 not defined

X7 with AFHDS3 defined
String issue. This is correct
![image](https://user-images.githubusercontent.com/5167938/80360413-9af20d80-887f-11ea-9c4d-a96ac24d5e43.png)

This is not
![image](https://user-images.githubusercontent.com/5167938/80360493-b78e4580-887f-11ea-95f8-ce98198f3c1c.png)

Accepting choice force closes

Same goes for X9E

And apparently 480 too, can't test anything

qba667 commented:
@3djc the simulator crash is because some MCU specific code is compiled and used - but advised by you I was not touch anything that is not relevant to this commit. So fix your code first or let me do it. The text will be fixed, missing string termination.

qba667 committed:
AFHDS3 text fixed according to comment in PR

kilrah commented:
Making things work in the simu is very much relevant and part of it :)

qba667 commented:
@kilrah no problem, I have the fix already - but this was for long there for me not relevant because it is affecting other components as well. The file /taranis/telemetry_driver.cpp should not be included in compilation of target simu/simulator.
Fix will be committed in few minutes.

3djc commented:
Same goes for protocol. There is also a dead spot betwen range and frequency

qba667 committed:
Workaround for simulator and AFHDS3.

qba667 committed:
AFHDS3 texts improved - as discussed dinPR

qba667 commented:
All those files should be excluded from building simulator target.

taranis/haptic_driver.cpp
taranis/telemetry_driver.cpp
taranis/extmodule_driver.cpp
/taranis/backlight_driver.cpp
/taranis/trainer_driver.cpp
/taranis/intmodule_pulses_driver.cpp

We should define just empty functions for all hardware related functions, and keeps those that are doing hardware independent actions.
Then we can remove ifdefs from source. But this all is bigger change. 
Fixing it correctly only for single file does not make much sense to me, fixing it trigger also compliment that I am changing to much. @kilrah after we discuss it I can commit separated PR for those changes.

qba667 commented:
@3djc  spacing removed form text - this was caused by other bug in compiler this time- if setting text length to "\008" it was not calculated correctly for choices and text was cut, using hex representation "\x008" works just fine.

qba667 commented:
@3djc @kilrah for now I committed workaround allowing you testing it on simulator - but on longer run it should be cleaned up.

3djc commented:
> @3djc spacing removed form text - this was caused by other bug in compiler this time- if setting text length to "\008" it was not calculated correctly for choices and text was cut, using hex representation "\x008" works just fine.

Value is in Octal, so yes \008 wont work, should be \010

3djc commented:
Your power values do not follow our standard either, they need to be \0 terminated after last alphanumerical. Dead spot after range still present

![image](https://user-images.githubusercontent.com/5167938/80400603-53897280-88bb-11ea-9e63-e4aecca00ee6.png)

3djc commented on radio/src/gui/480x272/model_setup.cpp:
```diff
@@ -1076,6 +1090,11 @@ bool menuModelSetup(event_t event)
+#endif
+#if defined(AFHDS3)
+        else if (isModuleAFHDS3(moduleIdx)) {
+          lcdDrawTextAtIndex(MODEL_SETUP_4TH_COLUMN, y, STR_AFHDS3_PROTOCOLS, g_model.moduleData[moduleIdx].rfProtocol, (menuHorizontalPosition==1 ? attr : 0));
```
Maybe lcdNextPos + 20 is a nicer choice than MODEL_SETUP_4TH_COLUMN which is quite far right ?

3djc commented:
![image](https://user-images.githubusercontent.com/5167938/80402972-11623000-88bf-11ea-8b54-50be22deb29a.png)

3djc commented:
![image](https://user-images.githubusercontent.com/5167938/80403016-25a62d00-88bf-11ea-8dbe-4495601178e8.png)

qba667 committed:
Text fixed accordig to PR comments.

qba667 commented:
![obraz](https://user-images.githubusercontent.com/3341647/80421908-7e84be00-88dd-11ea-8816-5d6eed429909.png)
![obraz](https://user-images.githubusercontent.com/3341647/80421971-a2480400-88dd-11ea-9c9f-5fa59c4fdea5.png)
![obraz](https://user-images.githubusercontent.com/3341647/80422004-affd8980-88dd-11ea-8966-bbb4937b2cd8.png)

qba667 committed:
Afhds 3 text update. Model setup for 128x64 fixed.

qba667 committed:
Handling of not ready state improved for afhds3

3djc commented:
Much better :) I’ll review 212 hopefully tomorrow. I’ll leave pulse code for someone else to review, but I noted

sizeof(ExternalModulePulsesData)
2.3 X9E : : 616 bytes
With this PR : 1400 bytes

2.3 X10E : 1216 bytes
With this PR : 2680 bytes

qba667 committed:
Channel model for FRM302 improved, fail safe fixed.

qba667 commented:
@3djc I it looks to me like to big buffer allocated for timer based pulses data. I will cheek this one as well. This is drawback of using 32 bit timer to generate software USART signal, and lack of signal inversion configuration for HW USART. I will reduce the buffer to smallest possible.
We can reconfigure Timer setup/DMA setup so 32 timer will use 16 byte values in this case it will save a lot of buffer space. Reducing buffer size to maximum size of frame will also reduce it.
Actually worst case 64 bytes of data every byte may be up to 10 output changes while generating USART so 640 x 32 bits = 2560 bytes to send 64 byte buffer.

3djc commented:
![image](https://user-images.githubusercontent.com/5167938/80567834-2d1c2200-89f6-11ea-8b3b-bb740a89e0cb.png)

212 is looking ok, you might want to push subtype a bit right but that really minor. You still have a glitch on navigation, it takes two clicks to go from [Range] to frequency when only one should be needed, there is a glitch in menu definitions

qba667 committed:
Fixed channels number for afhds3.

qba667 committed:
Using 16 bits values even for 32 bit timer - reducing memory usage - afhds3.

qba667 committed:
FRM302 dynamic memory allocagtion removed.
Fixed max frame size for AFHDS3.
AFHDS3 frame generation now checks array bounds.

qba667 committed:
Fixed menu item count for FRM302.

qba667 commented:
@3djc menu bug fixed.
Also reduced memory usage for timer based pulses 320*2bytes for timer data.

qba667 committed:
Compilation fixed for FRM302

qba667 committed:
Fixed TRACE usage in release build.

kilrah commented:
Looks like receiver number can't be edited, is it intended?

Also one issue, switching module type to MULTI doesn't work without rebooting the radio

kilrah committed:
Merge branch '2.3' into FRM302

# Conflicts:
#	radio/src/translations/untranslated.h

qba667 committed:
Deiniting not inverted telemetry mode on telemetry init.

3djc commented:
- it needs to be off by default and proposed as a firmware option in companion
- companion gui needs to be added to allow the setup of this module in companion like any others

bsongis commented on radio/src/pulses/afhds3.cpp:
needed?

bsongis replied:
DUMP does the same if I am not mistaken

qba667 replied:
DUMP indeed is similar - function not used any-more - if needed DUMP combined with other TRACE can be used.

bsongis commented on radio/src/pulses/afhds3.cpp:
```diff
@@ -0,0 +1,661 @@
+/*
```
cosmetics: is it possible you check the indentation, new lines, spaces in this file?

qba667 replied:
Reformatted using settings: 
![obraz](https://user-images.githubusercontent.com/3341647/82154400-e3449000-986d-11ea-9ebb-632dc808a492.png)
Feel free to edit attached to this PR Eclipse style. Maybe we can publish it later as official style for the project?

3djc replied:
We have included a CLion style sheet in the repo, but sure, we could add another ide too

qba667 replied:
@3djc can you just list all settings?
I don't want to install next IDE to get those settings.
Eclipse and QT editor settings would make sense to me. CLion is not free software so it is not fitting in project at all.

3djc replied:
It is free for open source projects :)

bsongis replied:
This code style is not the one we use in opentx. 
For example if / else if / else is completely wrong

bsongis commented on radio/src/pulses/afhds3.h:
```diff
@@ -0,0 +1,523 @@
+  void clearCommandFifo();
+  inline uint32_t nextIndex(uint32_t idx)
+  {
+    return (idx + 1) & (sizeof(commandFifo)/sizeof(commandFifo[0]) - 1);
```
DIM

bsongis commented on radio/src/pulses/afhds3.h:
```diff
@@ -0,0 +1,523 @@
+  }
+  inline uint32_t prevIndex(uint32_t idx)
+  {
+    if(0) return (sizeof(commandFifo)/sizeof(commandFifo[0]) - 1);
```
needed?

qba667 replied:
Not used now - thus I have not spotted it. Fixed by checking idx against 0. Can be needed in feature.

bsongis commented on radio/src/pulses/afhds3.h:
```diff
@@ -0,0 +1,523 @@
+/*
```
code style to review in this whole file

qba667 replied:
Applied - please see attached eclipse style.
It seems it has small issues while handling marco definition around structs like PACKED.

bsongis commented on radio/src/pulses/pulses.cpp:
```diff
@@ -40,6 +41,52 @@ void ModuleState::startBind(BindInformation * destination, ModuleCallback bindCa
 #endif
 }
 
+void getModuleStatusString(uint8_t moduleIdx, char * statusText) {
```
newline before {

qba667 replied:
Fixed, formatting rules applied.

bsongis commented on radio/src/pulses/pulses.cpp:
```diff
@@ -40,6 +41,52 @@ void ModuleState::startBind(BindInformation * destination, ModuleCallback bindCa
+void getModuleSyncStatusString (uint8_t moduleIdx, char * statusText) {
+  *statusText = 0;
+#if defined(MULTIMODULE)
+  if(isModuleMultimodule(moduleIdx)) {
```
space between if and (

qba667 replied:
Fixed, formatting rules applied.

bsongis commented on radio/src/pulses/pulses.cpp:
```diff
@@ -40,6 +41,52 @@ void ModuleState::startBind(BindInformation * destination, ModuleCallback bindCa
+  }
+#endif
+}
+void getModuleSyncStatusString (uint8_t moduleIdx, char * statusText) {
```
empty line between 2 functions

qba667 replied:
Fixed, formatting rules applied.

qba667 commented:
[OpenTX_style.zip](https://github.com/opentx/opentx/files/4640637/OpenTX_style.zip)

@bsongis 
Please update documentation if you want avoid such formatting issues.
https://github.com/opentx/opentx/wiki/OpenTX-Code-Style-Guide

The only difference from documentation is

> the beginning curly brace goes on the next line

I will rework pointed topics but documentation must be updated.

qba667 committed:
Queue include removed.

qba667 committed:
Formatting applied according to PR discussion

qba667 committed:
Merge branch '2.3' into FRM302

qba667 committed:
Spacing fixed.

qba667 committed:
Spacing fixed.

qba667 committed:
Merge branch 'FRM302' of https://github.com/opentx/opentx into FRM302

qba667 committed:
Companion config of AFHDS3 added.

qba667 committed:
Fixed compilation issues for Companion.

qba667 committed:
Missing declaration for multi module added.

qba667 committed:
Merge branch '2.3' into FRM302

bsongis commented:
@qba667 I am sorry if the wiki page about the code style is not up to date, but it exists since 2017 and I think everything there is still valid. You didn't even read / follow it. So please remove this: "Please update documentation if you want avoid such formatting issues". It's just not serious.

bsongis commented on companion/src/firmwares/opentx/opentxeeprom.cpp:
```diff
@@ -2114,6 +2114,40 @@ class ModuleUnionField: public UnionField<unsigned int> {
+        return attr == PULSES_AFHDS3;
+      }
+
+      void beforeExport() override {}
```
needed?

qba667 replied:
Needed because pure virtual function will not compile on Travis build.

bsongis commented on companion/src/firmwares/opentx/opentxeeprom.cpp:
```diff
@@ -2114,6 +2114,40 @@ class ModuleUnionField: public UnionField<unsigned int> {
+
+      void beforeExport() override {}
+
+      void afterImport() override {}
```
needed?

qba667 replied:
Needed because pure virtual function will not compile on Travis build.

bsongis replied:
ok, I wonder why I made it pure virtual then

bsongis commented on companion/src/firmwares/opentx/opentxeeprom.cpp:
```diff
@@ -2221,13 +2257,19 @@ class ModuleField: public TransformedField {
         module.rfProtocol = module.multi.rfProtocol & 0x0F;
       }
+      else if (module.protocol == PULSES_AFHDS3) {
+        module.rfProtocol = module.subType;
```
We will merge rfProtocol and subType in 2.4
I suggest you use subType for this module

qba667 replied:
Done

bsongis commented on radio/src/gui/212x64/model_setup.cpp:
```diff
@@ -810,6 +825,11 @@ void menuModelSetup(event_t event)
           lcdDrawTextAtIndex(lcdNextPos + 3, y, STR_R9M_REGION, g_model.moduleData[EXTERNAL_MODULE].subType, (menuHorizontalPosition==1 ? attr : 0));
+#if defined(AFHDS3)
+        else if (isModuleAFHDS3(EXTERNAL_MODULE)) {
+          lcdDrawTextAtIndex(lcdNextPos + 3, y, STR_AFHDS3_PROTOCOLS, g_model.moduleData[EXTERNAL_MODULE].rfProtocol, (menuHorizontalPosition==1 ? attr : 0));
```
subType

qba667 replied:
Done

3djc commented:
Just had a look at the companion implementation, looks fine. Just a quick question, since I don't have any of those hardware, does "receiver" failsafe mode exist on those ?

bsongis commented on radio/src/pulses/modules_constants.h:
```diff
@@ -217,4 +218,13 @@ enum AntennaModes {
 };
 
+enum AfhdsModes {
+  AFHDS_MODE_FIRST,
```
perhaps use AFHDS_SUBTYPE_FIRST, AFHDS_SUBTYPE_PWM_IBUS ... AFHDS_SUBTYPE_LAST

qba667 replied:
Renamed as requested.

bsongis commented on radio/src/pulses/modules_helpers.h:
```diff
@@ -299,8 +299,30 @@ inline bool isModuleSBUS(uint8_t moduleIdx)
+  return g_model.moduleData[idx].type == MODULE_TYPE_AFHDS3;
+}
+
+inline bool isModulePxx2(uint8_t idx){
```
newline missing

qba667 replied:
Fixed

bsongis commented on radio/src/pulses/modules_helpers.h:
```diff
@@ -299,8 +299,30 @@ inline bool isModuleSBUS(uint8_t moduleIdx)
+}
+
+inline bool isModulePxx2(uint8_t idx){
+  if(idx != EXTERNAL_MODULE) return false;
```
newline before return false
space between if and (

qba667 replied:
Fixed

bsongis commented on radio/src/pulses/modules_helpers.h:
```diff
@@ -299,8 +299,30 @@ inline bool isModuleSBUS(uint8_t moduleIdx)
+  if(idx != EXTERNAL_MODULE) return false;
+#if defined(PXX2)
+  #if defined(HARDWARE_EXTERNAL_MODULE_SIZE_SML)
+     if(g_model.moduleData[idx].type == MODULE_TYPE_XJT_LITE_PXX2 ||
```
space between if and (

qba667 replied:
Fixed as before

bsongis commented on radio/src/pulses/modules_helpers.h:
```diff
@@ -299,8 +299,30 @@ inline bool isModuleSBUS(uint8_t moduleIdx)
+       return true;
+  #endif
+  #if defined(HARDWARE_EXTERNAL_MODULE_SIZE_SML)
+    if(g_model.moduleData[idx].type == MODULE_TYPE_R9M_LITE_PXX2)
```
space between if and (

qba667 replied:
Fixed as well

bsongis commented on radio/src/pulses/modules_helpers.h:
```diff
@@ -325,6 +347,9 @@ inline int8_t maxModuleChannels_M8(uint8_t moduleIdx)
       return 8; // always 16 channels in FCC / FLEX
     }
   }
+  else if(isModuleAFHDS3(moduleIdx)){
```
space between if and (

qba667 replied:
Fixed

bsongis commented on radio/src/pulses/modules_helpers.h:
```diff
@@ -543,6 +573,28 @@ inline void resetAccessAuthenticationCount()
 #endif
 }
 
+inline void resetAfhds3Options(uint8_t moduleIdx){
```
newline missing

qba667 replied:
Fixed

bsongis commented on radio/src/pulses/modules_helpers.h:
```diff
@@ -543,6 +573,28 @@ inline void resetAccessAuthenticationCount()
 }
 
+inline void resetAfhds3Options(uint8_t moduleIdx){
+  g_model.moduleData[moduleIdx].rfProtocol = 0;
```
auto & module = g_model.moduleData[moduleIdx];

qba667 replied:
Fixed as requested - compiler is optimising it anyhow.

bsongis commented on radio/src/pulses/modules_helpers.h:
```diff
@@ -299,8 +299,30 @@ inline bool isModuleSBUS(uint8_t moduleIdx)
+
 // order is the same as in enum Protocols in myeeprom.h (none, ppm, pxx, pxx2, dsm, crossfire, multi, r9m, r9m2, sbus)
-static const int8_t maxChannelsModules[] = { 0, 8, 8, 16, -2, 8, 4, 8, 16, 8}; // relative to 8!
+//qba667 count is not matching!
```
what does this comment mean?

qba667 replied:
Look at the code in  maxModuleChannels_M8
`return maxChannelsModules[g_model.moduleData[moduleIdx].type];`
It use module type defined as ModuleType
```
MODULE_TYPE_NONE
MODULE_TYPE_PPM,
MODULE_TYPE_XJT_PXX1,
MODULE_TYPE_ISRM_PXX2,
MODULE_TYPE_DSM2,
MODULE_TYPE_CROSSFIRE,
MODULE_TYPE_MULTIMODULE,
MODULE_TYPE_R9M_PXX1,
MODULE_TYPE_R9M_PXX2,
MODULE_TYPE_R9M_LITE_PXX1,
MODULE_TYPE_R9M_LITE_PXX2,
MODULE_TYPE_R9M_LITE_PRO_PXX1,    // Doesn't exist
MODULE_TYPE_R9M_LITE_PRO_PXX2,
MODULE_TYPE_SBUS,
MODULE_TYPE_XJT_LITE_PXX2,
MODULE_TYPE_AFHDS3,
```
The array has to few items to reflect all module types - this should be fixed but as instructed I am not including any necessary changes.

bsongis commented on radio/src/pulses/afhds3.cpp:
```diff
@@ -0,0 +1,790 @@
+#include "../debug.h"
+#include "../definitions.h"
+#include <cstdio>
+#define FAILSAFE_HOLD 1
```
newline here

qba667 replied:
Fixed

bsongis commented on radio/src/pulses/afhds3.cpp:
```diff
@@ -0,0 +1,790 @@
+#define FAILSAFE_CUSTOM 2
+#define MAX_RETRIES_AFHDS3 5
+
+extern int32_t getChannelValue(uint8_t channel);
```
would be better in a .h if you introduce a function for it
I think it could be even inlined

qba667 replied:
Declaration of this function can be done in afhds3.h however it belongs to global namespace and is defined in pulses.cpp - doing it in pulses.h will case circular referencing.
For now I just declare it in afhds3.h.

bsongis commented on radio/src/targets/taranis/CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@ option(SHUTDOWN_CONFIRMATION "Shutdown confirmation" OFF)
 option(LCD_DUAL_BUFFER "Dual LCD Buffer" OFF)
 option(PXX1 "PXX1 protocol support" ON)
 option(PXX2 "PXX2 protocol support" OFF)
+option(AFHDS3 "AFHDS3 TX Module" ON)
```
This needs to be discussed, do we want it supported as default? It uses a lot of RAM (am I right?). So it means more users not happy because some Lua scripts won't run anymore.

I am OK for Horus. We don't have this issue. But for Taranis I really want to be sure that this PR won't bring problems.

qba667 replied:
Just for your reference size of ExternalModulePulsesData with AFHDS enabled for X9D - 864 bytes without 616 bytes. Lets disable it on F2.

qba667 replied:
Implementation in Companion added, fwoptions.py filled with needed options.
Is there anything more to be done?

bsongis commented:
I don't see an option in Companion? I see no other solution for Taranis users (provided we agree that this protocol uses more RAM for nothing for people who don't have this module)

qba667 committed:
Using subType instead of rfProtocol.

qba667 commented:
> n those ?

Yes it does, I send 2 of them to Andre - if I will get more I will provide them to you as well.

qba667 commented:
> Just had a look at the companion implementation, looks fine. Just a quick question, since I don't have any of those hardware, does "receiver" failsafe mode exist on those ?

Yes it does, I provided 2 of modules to Andre, as I will get more I will also provide them to you.

3djc commented:
> > Just had a look at the companion implementation, looks fine. Just a quick question, since I don't have any of those hardware, does "receiver" failsafe mode exist on those ?
> 
> Yes it does, I provided 2 of modules to Andre, as I will get more I will also provide them to you.

Cool, no it’s ok like that, I do not need one, just wanted to confirm there is a way to set those failsafe on receiver only without radio been involved. Thx

qba667 committed:
Using AFHDS_SUBTYPE as requested.

qba667 committed:
New lines in if fixed

qba667 committed:
The function getChannelValue declared in afhds3.
Formatting impoved.

qba667 committed:
AFHDS3 disabled default for F2 radios.

kilrah commented:
I have not found such a possibility, the manual for the receiver even says failsafe has to be set from the radio.

qba667 committed:
Afhds3 build options added.

qba667 commented:
> 
> 
> I have not found such a possibility, the manual for the receiver even says failsafe has to be set from the radio.

This is not right - we need to mode and values that can be preconfigured. If we take model definition from companion failsafe values will be just set during operation.

qba667 commented:
@bsongis Topic to be clarified is formatting:
From documentation:
> Block statements
>     block statements begin with an opening curly brace on the same line
> 

And just if statement example instead of block statement - it is not saying anything about if else syntax.

Most of code is written using following rules:

The if and else keywords belong on separate lines. There should be a space between the if and the open parenthesis, and between the close parenthesis and the curly brace (if any), but no space between the parentheses and the condition.

But you are requesting also new line between close parenthesis and the curly brace. 
This must be clearly defined, I know you are busy with other tasks, thus I can write verbal description of code formatting for the project, but you will need still to review it, just message me, we will find right way.

3djc commented:
> > I have not found such a possibility, the manual for the receiver even says failsafe has to be set from the radio.
> 
> This is not right - we need to mode and values that can be preconfigured. If we take model definition from companion failsafe values will be just set during operation.

"Receiver" failsafe in failsafe mode means the radio does nothing, but you tell OpenTX that a failsafe has been set using a manual process on the receiver. If there is no such manual process for this generation of receivers, then this specific value must be disabled I believe

qba667 commented:
> 
> 
> > > I have not found such a possibility, the manual for the receiver even says failsafe has to be set from the radio.
> > 
> > 
> > This is not right - we need to mode and values that can be preconfigured. If we take model definition from companion failsafe values will be just set during operation.
> 
> "Receiver" failsafe in failsafe mode means the radio does nothing, but you tell OpenTX that a failsafe has been set using a manual process on the receiver. If there is no such manual process for this generation of receivers, then this specific value must be disabled I believe


Ok got it - indeed in FW implementation I skipped it - here it need to be removed as well.

qba667 committed:
Reciever mode for failsafe removed for AFHDS3.

qba667 committed:
If else formatting updated.

qba667 commented:
@3djc @kilrah option removed for Afhds3 - is this only one not supporting this mode?

3djc commented:
> @3djc @kilrah option removed for Afhds3 - is this only one not supporting this mode?

I think so, at least for the modules we natively support. I guess any module offering a PPM mode has to offer that "receiveir only" option. But probably AFHDS3 is not the last one tho, seems to me  like most new ones will not have it

qba667 committed:
Spacing removed inside namespace.

qba667 committed:
More formatting fixes.

qba667 committed:
Merge branch '2.3' into FRM302

qba667 commented:
Development documentation updated, conflicts resolved.

qba667 committed:
Merge branch '2.3' into FRM302

qba667 committed:
Merge branch '2.3' of https://github.com/opentx/opentx into FRM302

bsongis committed:
Merge with latest 2.3

koakumaping commented:
no x12s build?I want to have a test.

kilrah commented:
Use nightlies...

koakumaping commented:
I tried. but failed（opentx-x12s-internalaccess-lua-ppmus-en-2.3.10N320）
![IMG_0841](https://user-images.githubusercontent.com/4099388/93012537-e9f0d300-f5d3-11ea-9744-02a9b37d085f.JPG)

3djc commented:
From what we heard, it needs a special firmware on the module to work with OpenTX, please contact Flysky to get this firmware

koakumaping commented:
> From what we heard, it needs a special firmware on the module to work with OpenTX, please contact Flysky to get this firmware

Thanks.I will send a email to flysky.

SUPANOOB-ZA commented:
Good evening 
I have been testing the FRM302 on the Flysky Nirvana and have encountered two issues, maybe you know of a solution that is being developed?

1) When using internal power mode, the module rapidly drains the Nirvana battery. The only workable solution is to use the external power option even on a low power setting such as 100mw. This does not seem logical as this should not drain 4.2v down to 3.7v on 3x 3000mah samsung lion batteries.

2) The module does not store previously bound receivers in "memory" on the Nirvana, which means I have to re-bind each time I fly a new craft with the module. Is this normal?

Thanks

qba667 commented:
@SUPANOOB-ZA for how long the batteries last?
The problem of nv14 is that batteries are never fully charged and there is internal drop of voltage causing we are not able to use full potential of batteries. Can be easily fixed. The issue with not storing bound model data I need to check. I will back to you soon.

@koakumaping have you got the right firmware? I can send you one.

SUPANOOB-ZA commented:
@qba667 
Re. Batteries
The Nirvana batteries can last for a whole day of racing previously so that's maybe 10 to 20 flights without a problem using the normal FS-X8B receivers in my quad. I have never experienced an issue like this before. But for now I have a work around by using the external battery. I'm glad you say it can easily be fixed.

Re. Models and binding
I hope you have some luck here. My gut feeling is that when the receiver is bound, it is not being stored in a unique memory allocation per model, but rather a single memory allocation for " bound receivers to external module " ? I'm not a developer, just guessing here.

All the best, thank you for all that you do! The heroes of the hobby!

qba667 commented:
@SUPANOOB-ZA 
Here you have How To allowing solving you power issues.
The model ID is not returned from module, so we can not store it, but FlySky promised to add this feature in next firmware update.

SUPANOOB-ZA commented:
> @SUPANOOB-ZA
> Here you have How To allowing solving you power issues.
> The model ID is not returned from module, so we can not store it, but FlySky promised to add this feature in next firmware update.

Hi @qba667 
For Power issue,
Was there meant to be a link on your first line regarding power?

For Models,
I hope Flysky does this, it is a terrible issue.

Thanks again

qba667 commented:
@SUPANOOB-ZA 
here you have 
https://docs.google.com/document/u/1/d/e/2PACX-1vRy9U54N_76afuX7yOfTy7nqqF62sL8To3QSVMR0jCer2gfv5SF3BVrcI_PaiOqJ6d9yqaJ7veFAzQn/pub?urp=gmail_link&gxids=7628

vandusinae commented:
Hello, I try to run a stock FRM302 with the current OpenTX version 2.3.10-OTX (x9d+) and can't control it in serial mode - status: not ready - ppm works. Does it need a special firmware to work with OpenTX? I wrote to Flysky as suggested in this thread. Maybe someone has this firmware and can share it or a tutorial. Thanks

SUPANOOB-ZA commented:
> Hello, I try to run a stock FRM302 with the current OpenTX version 2.3.10-OTX (x9d+) and can't control it in serial mode - status: not ready - ppm works. Does it need a special firmware to work with OpenTX? I wrote to Flysky as suggested in this thread. Maybe someone has this firmware and can share it or a tutorial. Thanks

I believe that there is a custom firmware for radiomaster that supports it but its not widely available. 
Ref. https://www.youtube.com/c/PawełSpychalski

But the Nirvana has received updated firmware from Flysky via their own firmware update application. 
Ref. https://youtu.be/w8by7oW0m7c

vandusinae commented:
Thank you! I'll write to Pawel. Cheers

qba667 commented:
@vandusinae @koakumaping here you have
[Update firmware FRM302 115200.zip](https://github.com/opentx/opentx/files/5356324/Update.firmware.FRM302.115200.zip)

vandusinae commented:
Thank you!!

vandusinae commented:
Still dosen't work on a x9d+, Module status: not ready, PPM (tx) works

qba667 commented:
@vandusinae so something must be changed in OpenTX in between. Will try to verify it as soon as possible. Can you contact me directly so we can check few things?

vandusinae commented:
Thanks! I tested another x9d+ and a X-Lite S and reflashed your firmware. But still the same result.

qba667 commented:
@vandusinae  Let me provide you another firmware with different software USART timings. 
[firmware_x9d+.zip](https://github.com/opentx/opentx/files/5358899/firmware_x9d%2B.zip)
Can you please check this one?

vandusinae commented:
I flashed it to two x9d+, but still Module status: not ready, nevertheless I tried to rebind in serial mode, but nothing happens. ppm works

qba667 commented:
@vandusinae, thanks for testing, the right fix is on the way!

vandusinae commented:
Great!! Thank you!

SUPANOOB-ZA commented:
**EDITED & UDPATED**

I did some testing regarding the FlySky Nirvana and FRM302 transmitter. To try resolve an internal battery voltage drain issue.
FlySky Nirvana, FRM302 and Samsung Lithium Ion Batteries

Initially I thought the new batteries solved the issue, however once I was connected to the receiver, the NV14 experienced the voltage drainage issue, temporarily from 4.2v to 3.6v. Voltage restores back to 4.2v once the external module is removed / not in us. This is most likely because I have the HW v1.4 main board. 

I've given up and ordered the Radiomaster tx16s with hall sensor gimbals with Cross Fire micro tx v2 combo. I will still test the FRM302 and hopefully I can find an application to use it for. The smaller antennas definitely are a plus compared to XF.

qba667 commented:
@vandusinae can we start from checking new version of module firmware?
[Update _fix firmware FRM302_115200.zip](https://github.com/opentx/opentx/files/5361030/Update._fix.firmware.FRM302_115200.zip)

vandusinae commented:
It seems to work now. I can control the module via serial mode. I bound the receiver with telemetry. There is a buzzing noise in the radio speaker, when switching the output power from 100 to 500. Similar to Crossfire on high output power. That means switching between the output power levels work. I'll try a test flight in the next days and test the failsafe options. Many thanks!!! Cheers
https://youtu.be/1BYQk04Y29o

vandusinae commented:
The module is also recognized by the X-Lite S, binding works, but I can't change the protocol. The cursor jumps from mode (AFHDS3) to Ch Range if I try to get to Proto. This has to be probably fixed in OpenTX. 

Mode
Proto
Type
Status Power src.
Ch. Range

qba667 commented:
@vandusinae I have already committed a pull request fixing the problem. I can also build for you test firmware.

vandusinae commented:
yes please, I don't have an OpenTx dev environment installed yet

vandusinae commented:
Compiled it and it works, Proto can be selected on a X-Lite.

OT: is there a list of targets and modules? ; e.g. xlites, afhds3, ... I couldn't find one via google and had to look into the code

SUPANOOB-ZA commented:
Is there opentx for the radiomaster tx16s that supports rhe FRM302? 

Thanks in advance

3djc commented:
2.3.10

SUPANOOB-ZA commented:
> > From what we heard, it needs a special firmware on the module to work with OpenTX, please contact Flysky to get this firmware
> 
> Thanks.I will send a email to flysky.

Please can you share the firmware here. I have now bought a Radiomaster TX16S using opentx 2.3.10 and would like to get my FRM302 working :) 


Thanks

vandusinae commented:
OpenTX 2.3.13 AHDS3 support problem, 

radio: x9d+, firmware file: opentx-x9d+-afhds3-lua-noheli-sqt5font-en-2.31.3-otx

Eventhough I selected the AFHDS3 module in the Companion frontend and upgraded the firmware, the AFHDS3 module doesn't appear under general settings 7/7, firmware options and it can't  be selected as External RF. It worked without a problem in the previous 2.3.10-otx version . I switched back to that version and it works again. Here's another FRM302 related issue: https://github.com/opentx/opentx/issues/8606 maybe they are connected. Is there something I can test?

Cheers
